ipatests/test_ipalib/test_filestore.py

```diff
@@ -0,0 +1,985 @@
+Module provides unit tests to verify that the FileStore code works.
+"""
+
+from __future__ import absolute_import
+from hashlib import sha256
+from ipaplatform.paths import paths
+
+try:
+    # install subpackages are not shipped with wheels
+    from ipalib.install import sysrestore
+    from ipalib.install.sysrestore import SYSRESTORE_INDEXFILE as INDEX
+    from ipalib.install.sysrestore import SYSRESTORE_MAX_INDEX as MAX_INDEX
+    from ipalib.install.sysrestore import SYSRESTORE_PATH_INDEX as PATH_INDEX
+    from ipalib.install.sysrestore import SYSRESTORE_SEP as SEP
+except ImportError:
+    INDEX = MAX_INDEX = PATH_INDEX = SEP = None
+
+import collections
+import filecmp
+import os
+import pytest
+import shutil
+import six
+import uuid
+
+# pylint: disable=import-error
+if six.PY3:
+    # The SafeConfigParser class has been renamed to ConfigParser in Py3
+    from configparser import ConfigParser as SafeConfigParser
+    from configparser import MissingSectionHeaderError
+else:
+    from ConfigParser import SafeConfigParser
+    from ConfigParser import MissingSectionHeaderError
+# pylint: enable=import-error
+
+# install subpackages are not shipped with wheels
+pytestmark = [pytest.mark.needs_install_package, pytest.mark.tier0]
+
+STORE_ROOT_DIR = os.path.join(paths.TMP, "sysrestore")
+STORE_DIR = os.path.join(STORE_ROOT_DIR, str(uuid.uuid4()))
+SOURCE_DIR = os.path.join(paths.TMP, "sourcedir")
+
+
+# helpers
+def backuped_name(path):
+    """
+    Return an expected internal name of a file being backuped to filestore
+    """
+    backupfile = os.path.basename(path)
+    with open(path, 'rb') as f:
+        cont_hash = sha256(f.read()).hexdigest()
+    return "{hexhash}-{bcppath}".format(
+        hexhash=cont_hash, bcppath=backupfile
+    )
+
+
+def mkdir_store():
+    try:
+        shutil.rmtree(STORE_DIR)
+    except FileNotFoundError:
+        pass
+    os.makedirs(STORE_DIR)
+    return STORE_DIR
+
+
+def mkdir_source():
+    try:
+        shutil.rmtree(SOURCE_DIR)
+    except FileNotFoundError:
+        pass
+    os.makedirs(SOURCE_DIR)
+    return SOURCE_DIR
+
+
+def mkfile_source(sourcefile="sourcepath", content="content"):
+    sourcepath = os.path.join(SOURCE_DIR, sourcefile)
+    with open(sourcepath, "w") as f:
+        f.write(content)
+    return sourcepath
+
+
+def setup_module(module):
+    if os.path.isfile(os.path.join(paths.TMP, INDEX)):
+        pytest.fail("Default filestore index file is already present. "
+                    "These tests may break IPA installation.")
+    try:
+        shutil.rmtree(STORE_ROOT_DIR)
+    except FileNotFoundError:
+        pass
+
+
+def setup_function(function):
+    try:
+        os.unlink(os.path.join(paths.TMP, INDEX))
+    except FileNotFoundError:
+        pass
+    mkdir_store()
+    mkdir_source()
+
+
+def idf(val):
+    if not val:
+        return "@default"
+    else:
+        return "@" + val
+
+
+@pytest.fixture(params=["", os.path.join(SOURCE_DIR, "newpath")], ids=idf)
```

Comment:
Why are there so many fixtures and what do these fixtures actually do? Couldn't you just use parametrized to archive the same goal with less code? I'm worried that nobody will understand the meaning of the code in a couple of years.

Reply:
I dropped almost all fixtures to simplify the code.